### PR TITLE
py-knights-fighting

### PR DESCRIPTION
**Pull request by muriakula**

### CODE REVIEW

File: app/stats/armor.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+    def __init__(self, part: str, protection: int) -> None:
+        self.part = part
+        self.protection = protection
+
+    @staticmethod
+    def armour_registration(knights_armour: list[dict]) -> list["Armour"]:
+        registered_armour = []
+        for part in knights_armour:
+            registered_armour.append(Armour(part["part"], part["protection"]))
+        return registered_armour
```
Comment (Bohdan-Salamakha) [inline app/stats/armor.py:11]: use list comprehension for it

File: app/stats/knights.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    knights = {}
+
+    def __init__(
+            self,
+            name: str,
+            power: int,
+            hp: int,
+            armour: list["Armour"] = None,
+            weapon: "Weapon" = None,
+            potion: "Potion" = None
```
Comment (Bohdan-Salamakha) [inline app/stats/knights.py:16]: use typehint like: Optional[YOUR_VALUE_TYPE] = None
for example: list that contains string elements and can be None has typehint Optional[list[str]]

File: app/stats/potion.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+            hp: int = 0,
+            power: int = 0
+    ) -> None:
+        self.name = name
+        self.protection = protection
+        self.hp = hp
+        self.power = power
+
+    @staticmethod
+    def potion_registration(knights_potion: dict = None) -> "Potion" or None:
```
Comment (Bohdan-Salamakha) [inline app/stats/potion.py:15]: u can't use "or" in the typehint like "Potion" or None; use Union[..., ...] or | (Pipe)

use typehint like: Optional[YOUR_VALUE_TYPE] = None
for example: list that contains string elements and can be None has typehint Optional[list[str]]

File: app/stats/potion.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        if knights_potion is not None:
+            effect = knights_potion["effect"]
+            protection = 0
+            if "protection" in effect:
+                protection = effect["protection"]
```
Comment (Bohdan-Salamakha) [inline app/stats/potion.py:20]: use ternary operator for this

File: app/stats/potion.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+            hp = 0
+            if "hp" in effect:
+                hp = effect["hp"]
```
Comment (Bohdan-Salamakha) [inline app/stats/potion.py:23]: use ternary operator for this

File: app/stats/potion.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+            power = 0
+            if "power" in effect:
+                power = effect["power"]
```
Comment (Bohdan-Salamakha) [inline app/stats/potion.py:26]: use ternary operator for this

File: app/stats/potion.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+            return Potion(
+                name=knights_potion["name"],
+                protection=protection,
+                hp=hp,
+                power=power
+            )
+        else:
```
Comment (Bohdan-Salamakha) [inline app/stats/potion.py:33]: remove redundant "else"

File: app/stats/weapon.py [hunk 8]
```diff
@@ -0,0 +1,8 @@
+class Weapon:
+    def __init__(self, name: str, power: str) -> None:
+        self.name = name
+        self.power = power
+
+    @staticmethod
+    def weapon_registration(knights_weapons: dir) -> "Weapon":
```
Comment (Bohdan-Salamakha) [inline app/stats/weapon.py:7]: invalid typehint "dir"

**Review by Bohdan-Salamakha — CHANGES_REQUESTED** [review]
v1

File: app/stats/knights.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
+from typing import Optional
+from app.stats.armor import Armour
+from app.stats.weapon import Weapon
+from app.stats.potion import Potion
+
+
+class Knights:
+    knights = {}
```
Comment (ArsenPidhoretskyi) [inline app/stats/knights.py:8]: Why you need to save all instances inside class?

File: app/stats/knights.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        self.protection = 0
+        self.name = name
+        self.power = power
+        self.hp = hp
+        self.armour = armour
+        self.weapon = weapon
+        self.potion = potion
+        Knights.knights[name] = self
+
+    @staticmethod
```
Comment (ArsenPidhoretskyi) [inline app/stats/knights.py:28]: It must be class method

File: app/stats/knights.py [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    def registration(knights: dict) -> "Knights":
```
Comment (ArsenPidhoretskyi) [inline app/stats/knights.py:29]: It returns None. But must return dict/list of knights

File: app/stats/knights.py [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    def battle_check(self) -> None:
+        if self.hp <= 0:
+            self.hp = 0
+
+    @staticmethod
+    def battle_preparations(knights: list["Knights"]) -> None:
+        for knight in knights:
+            knight.protection = Knights.apply_armour(knight)
+            knight.power = Knights.apply_weapon(knight)
+            Knights.apply_potion(knight)
```
Comment (ArsenPidhoretskyi) [inline app/stats/knights.py:64]: You can call method to instance. No need to write Knight.

File: app/stats/potion.py [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+            protection: int = 0,
+            hp: int = 0,
+            power: int = 0
+    ) -> None:
+        self.name = name
+        self.protection = protection
+        self.hp = hp
+        self.power = power
+
+    @staticmethod
```
Comment (ArsenPidhoretskyi) [inline app/stats/potion.py:17]: Use class method

File: app/stats/weapon.py [hunk 14]
```diff
@@ -0,0 +1,8 @@
+class Weapon:
+    def __init__(self, name: str, power: str) -> None:
+        self.name = name
+        self.power = power
+
+    @staticmethod
```
Comment (ArsenPidhoretskyi) [inline app/stats/weapon.py:6]: The same. Class method

**Review by ArsenPidhoretskyi — CHANGES_REQUESTED** [review]

File: app/stats/knights.py (hunk 9 above)
Comment (muriakula) [reply app/stats/knights.py:8]: this is needed to assign variables to the names of knights in main.py in lines 96-99

Comment (ArsenPidhoretskyi) [reply app/stats/knights.py:8]: But you can return it from your method. No need to save it into class

File: app/main.py [hunk 15]
```diff
@@ -86,129 +89,45 @@
...
     # 2 Arthur vs Red Knight:
-    arthur["hp"] -= red_knight["power"] - arthur["protection"]
-    red_knight["hp"] -= arthur["power"] - red_knight["protection"]
+    arthur.hp -= red_knight.power - arthur.protection
+    red_knight.hp -= arthur.power - red_knight.protection
 
     # check if someone fell in battle
-    if arthur["hp"] <= 0:
-        arthur["hp"] = 0
+    Knights.battle_check(arthur)
```
Comment (ArsenPidhoretskyi) [inline app/main.py:120]: Read how you can call methods to instances. Because it is not properly written code for now

File: app/stats/armor.py [hunk 16]
```diff
@@ -0,0 +1,10 @@
+class Armour:
+    def __init__(self, part: str, protection: int) -> None:
+        self.part = part
+        self.protection = protection
+
+    @staticmethod
```
Comment (ArsenPidhoretskyi) [inline app/stats/armor.py:6]: class methods also

File: app/stats/knights.py [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+        return [cls(
+            name=stat["name"],
+            power=stat["power"],
+            hp=stat["hp"],
+            armour=Armour.armour_registration(stat["armour"]),
+            weapon=Weapon.weapon_registration(stat["weapon"]),
+            potion=Potion.potion_registration(stat["potion"]))
+            for knight, stat in knights.items()]
+
+    def apply_armour(self) -> int:
```
Comment (ArsenPidhoretskyi) [inline app/stats/knights.py:39]: No need to return 
The same in apply_weapon

**Review by ArsenPidhoretskyi — CHANGES_REQUESTED** [review]

File: app/stats/knights.py (hunk 17 above)
Comment (muriakula) [reply app/stats/knights.py:39]: apply_potion and BATTLE part don`t work without return in apply_armour and apply_weapon. :(
How can I change it?

**Review by Bohdan-Salamakha — APPROVED** [review]

**Pull request closed without merging**